### ISSUE TITLE
CycleGAN in PyTorch Lightning

### ISSUE THREAD

**Issue by Adi-iitd** [PyTorchLightning/pytorch-lightning-bolts#397]
## 🚀 Feature
CycleGAN in PyTorch Lightning

### Motivation

The authors of this [paper](https://arxiv.org/pdf/1703.10593.pdf) proposed an **unpaired** image to image translation algorithm using the well-known **GAN** framework. Obtaining input-output pairs for graphics tasks like artistic stylization can be difficult since the desired output is highly complex, and typically requires artistic authoring. For many tasks, like object transfiguration (e.g., zebra <-> horse), the desired output is not even well-defined. So, this algorithm comes in very handy in such situations. Nowadays, researchers use CycleGAN even in the **Compressive Sensing** and **BioMedical** domain.

### Pitch

I already ported my code of CycleGAN from PyTorch to PyTorch Lightning. I want to add this in the PyTorch Lightning Bolts repo, but I still need help from your side to check the robustness of my code. You can find the code on my [repo](https://github.com/Adi-iitd/AI-Art/blob/master/src/CycleGAN/CycleGAN-PL.py).

### Alternatives

I couldn't find CycleGAN implementation in PyTorch Lightning. So, adding it would be of great help who are just starting in GAN research or want to use PyTorch Lightning to do advanced research in the GAN domain.

### Additional context

I am attaching a few results that I obtained to build some credibilty.
<img width="776" alt="Screenshot 2020-11-22 at 5 23 48 PM" src="https://user-images.githubusercontent.com/41862477/99975020-b1d1f180-2dc7-11eb-9b3c-4ae52d35b18d.png">
<img width="774" alt="Screenshot 2020-11-22 at 5 43 40 PM" src="https://user-images.githubusercontent.com/41862477/99975029-b5657880-2dc7-11eb-94f7-2698b0537a6c.png">
<img width="778" alt="Screenshot 2020-11-22 at 5 44 28 PM" src="https://user-images.githubusercontent.com/41862477/99975035-b5fe0f00-2dc7-11eb-81fa-cc96866df2c3.png">
<img width="775" alt="Screenshot 2020-11-22 at 5 44 44 PM" src="https://user-images.githubusercontent.com/41862477/99975038-b696a580-2dc7-11eb-8e84-393346a9e97b.png">
<img width="775" alt="Screenshot 2020-11-22 at 6 17 16 PM" src="https://user-images.githubusercontent.com/41862477/99975322-168d4c00-2dc8-11eb-92a6-d215d10d3459.png">
<img width="775" alt="Screenshot 2020-11-22 at 7 09 01 PM" src="https://user-images.githubusercontent.com/41862477/99975334-19883c80-2dc8-11eb-9299-0efd914ede11.png">

**Comment by ananyahjha93** [comment]
@Adi-iitd this seems like a good addition to the library. Feel free to create a PR for this. 

I think the code needs some minor changes but overall looks good. Do you have saved weights for the model? Or loss plots? We are encouraging saved checkpoints and loss plots (and images like you posted) to be added to alongside implementations to build credibility for the correctness of code.

**Comment by Adi-iitd** [comment]
@ananyahjha93 Yes, I do have the loss plots as well as the pre-trained weights. Attaching the TensorBorad loss logs here, kindly let me know how to share the saved checkpoints. I will open a PR for the same.

<img width="918" alt="Screenshot 2020-11-24 at 9 36 26 PM" src="https://user-images.githubusercontent.com/41862477/100119983-799ce280-2e9d-11eb-956e-f6ef6f6a05b1.png">
<img width="917" alt="Screenshot 2020-11-24 at 9 36 37 PM" src="https://user-images.githubusercontent.com/41862477/100119996-7bff3c80-2e9d-11eb-88f8-f9a0c1bb2ac4.png">
<img width="920" alt="Screenshot 2020-11-24 at 9 36 50 PM" src="https://user-images.githubusercontent.com/41862477/100120002-7c97d300-2e9d-11eb-911e-c08154295ca6.png">
<img width="917" alt="Screenshot 2020-11-24 at 9 37 01 PM" src="https://user-images.githubusercontent.com/41862477/100120003-7d306980-2e9d-11eb-90e1-2595a47523b3.png">
<img width="918" alt="Screenshot 2020-11-24 at 9 37 14 PM" src="https://user-images.githubusercontent.com/41862477/100120007-7dc90000-2e9d-11eb-9932-6285251dbade.png">

**Comment by akihironitta** [comment]
@Adi-iitd Sorry for the delay.

> Do you have saved weights for the model? Or loss plots? We are encouraging saved checkpoints and loss plots (and images like you posted) to be added to alongside implementations to build credibility for the correctness of code.

@ananyahjha93 @Borda I'd love to have this in Bolts, but I'm not aware of how to store pretrained weights in Bolts. Do we currently have guidance on this for contributors? Is it something like this?:
https://github.com/PyTorchLightning/pytorch-lightning-bolts/blob/99232c32602165410779c494b6bf492c2d291a71/pl_bolts/utils/pretrained_weights.py#L5-L6

**Comment by Borda** [comment]
yep would be nice to have, atm we have custom/manual model uploading... :rabbit:
